src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,476 @@
+  ++id_counter;
+  const code_declt declaration(symbol_exprt(var_id, type));
+  return declaration;
+}
+
+exprt symbol_analyzert::process_any_pointer_target(
+  const symbol_exprt &symbol,
+  const std::string memory_location,
+  const source_locationt &location)
+{
+  auto it = values.find(memory_location);
+  if(it == values.end())
+  {
+    const place_holder_exprt recursion_breaker(memory_location);
+    values[memory_location] = recursion_breaker;
+
+    const typet target_type = ns.follow(symbol.type().subtype());
+    const code_declt target_object = get_pointer_target(
+      id2string(symbol.get_identifier()), target_type, location);
+    generated_code.add(target_object);
+    values[memory_location] = target_object.symbol();
+    return target_object.symbol();
+  }
+  else
+  {
+    return it->second;
+  }
+}
+
+code_declt symbol_analyzert::get_pointer_target(
+  const std::string pointer_name,
+  const typet &type,
+  const source_locationt &location)
+{
+  code_declt declaration = declare_instance(type);
+
+  const std::string deref_pointer = "(*" + pointer_name + ")";
+  const exprt target_expr = fill_expr_with_values(
+    deref_pointer, zero_initializer(type, location, ns), type, location);
+
+  declaration.operands().resize(2);
+  declaration.op1() = target_expr;
+  return declaration;
+}
+
+exprt symbol_analyzert::fill_array_with_values(
+  const std::string &symbol_id,
+  exprt &array,
+  const typet &type,
+  const source_locationt &location)
+{
+  for(size_t i = 0; i < array.operands().size(); ++i)
+  {
+    const std::string cell_access = symbol_id + "[" + std::to_string(i) + "]";
+    const typet target_type = ns.follow(array.operands()[i].type());
+    if(is_c_char_pointer(target_type))
+    {
+      const std::string memory_location = gdb_api.get_memory(cell_access);
+      const symbol_exprt char_ptr(cell_access, target_type);
+      array.operands()[i] =
+        declare_and_initalize_char_ptr(char_ptr, memory_location, location);
+    }
+    else
+    {
+      array.operands()[i] = fill_expr_with_values(
+        cell_access, array.operands()[i], target_type, location);
+    }
+  }
+  return array;
+}
+
+exprt symbol_analyzert::fill_expr_with_values(
+  const std::string &symbol_id,
+  exprt expr,
+  const typet &type,
+  const source_locationt &location)
+{
+  if(expr.is_constant() && (is_c_int_derivate(type) || is_c_char(type)))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    const mp_integer int_rep = string2integer(value, DECIMAL_SYSTEM);
+    return from_integer(int_rep, type);
+  }
+  else if(expr.is_constant() && is_c_bool(type))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return from_c_boolean_value(value, type);
+  }
+  else if(expr.is_constant() && is_c_enum(expr.type()))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return convert_memeber_name_to_enum_value(
+      value, to_c_enum_type(expr.type()));
+  }
+  else if(is_struct(type))
```

Comment:
Should better be `else if(type.id() == ID_struct || type.id() == ID_struct_tag)`

Reply:
Done.